### ISSUE TITLE
Inverse parameter transforms and association with priors

### ISSUE THREAD

**Issue by Balandat** [cornellius-gp/gpytorch#381]
#376 changed priors to always operate on the actual parameters (e.g noise vs. its log or inverse softplus value). In some cases we want to sample parameters from the priors, which means that we'll have to transform the samples back to the internal parameter representation. We currently have no easy way of doing this. In particular, given a model, this requires additional knowledge of the model specifics. 

'Im thinking about ways of making this easier to automate. In the most general case this is hard, because (i) if a prior is on a function of two parameters that typically won't be a bijection, and (ii) if we have multiple priors that involve the same parameter it's unclear how to enforce consistency.

Let's focus on the simple case and consider only those priors and parameters for which there is a one-to-one mapping between priors and parameters, with all mappings bijective. 

One simple way would be to allow to specify an "inverse closure" when registering a prior for a single parameter. For instance, for a prior on the module's `noise` parameter, you could imagine something like:
```
.register_prior("foo_prior", foo_prior, self.noise, lambda x: self.noise = x)
```
which would require a `@setter` for the `noise` to be defined. This of course in turn would require that the user pass in an `inv_param_transform` to the module (that could be made optional by just having some dictionary mapping common transformations to their inverse, e.g. `{torch.exp: torch.log, torch.nn.functional.softplus: softplus_inverse...}` and error out when an unknown param_transform were supplied.

Then sampling from the prior and setting the parameter values could be done like
```
for name, prior, _, inv_closure in self.named_priors():
    if inv_closure is not None:
        inv_closure(prior.sample())
```

Thoughts on this?

cc @bletham

**Comment by jacobrgardner** [comment]
I mentioned this briefly on the PR, but one other thing an inverse transform would let us do more easily is sunset log_* without it being breaking. We can already write a `@property` getter that raises a depreciation warning and outputs a log lengthscale, but with an inverse transform we could write a setter that does the same.

This would let us more easily take the plunge to just using softplus everywhere.

**Comment by Balandat** [comment]
Ok, let me cook something up for this then

**Comment by jacobrgardner** [comment]
Closed via #383
